### PR TITLE
Now the user can decide if using deepcopy on input data.

### PR DESCRIPTION
**Pull request by Tosa95**
I was working on an image processing pipeline on Raspberry Pi using pyungo, I found that the deepcopy operation that happens in the constructor of the Data class took a long time, in the neighborhood of 20ms.

That is due to the fact that input data was a complete frame (a numpy array) that is quite big. Removing the deepcopy operation made me save those 20ms, that is a total performance gain of 14% (my fps is around 8 without deepcopy and around 7 with deepcopy).

I consider this a big improvement. So I added the possibility of disabling the deepcopy behaviour (since it is obviously a bad idea to disable it by default) and a warning that suggests the user to disable the deepcopy if the copy time goes over the 5% of total compute time for the graph.

As a note, on a standard pc the deepcopy operation takes like 0ms, but since on less powerful hardware the difference is so strong I think it is a good idea to implement this little modification.

### CODE REVIEW

File: pyungo/core.py [hunk 1]
```diff
@@ -223,8 +224,9 @@ class Graph:
         ImportError will raise in case parallelism is chosen and `multiprocess`
             not installed
     """
+
     def __init__(self, inputs=None, outputs=None, parallel=False, pool_size=2,
-                 schema=None):
+                 schema=None, do_deepcopy=True):
```
Comment (Tosa95) [inline pyungo/core.py:229]: A parameter that can be used to disable the deepcopy behavior

File: pyungo/core.py [hunk 2]
```diff
@@ -356,7 +361,10 @@ def calculate(self, data):
             jsonschema.validate(instance=data, schema=self._schema)
         t1 = dt.datetime.utcnow()
         LOGGER.info('Starting calculation...')
-        self._data = Data(data)
+        dt1 = dt.datetime.utcnow()
```
Comment (Tosa95) [inline pyungo/core.py:364]: Measuring the time taken to instantiate the Data object, so we can suggest the user to disable deepcopying if such time is too high

File: pyungo/core.py [hunk 3]
```diff
@@ -404,5 +412,16 @@ def calculate(self, data):
...
         t2 = dt.datetime.utcnow()
-        LOGGER.info('Calculation finished in {}'.format(t2-t1))
+        total_compute_time = t2 - t1
+        LOGGER.info('Calculation finished in {}'.format(total_compute_time))
+
+        data_copy_perc = data_copy_time.total_seconds() / total_compute_time.total_seconds()
+
+        if data_copy_perc > COPY_TIME_MAX_PERCENTAGE:
+            LOGGER.warning(
+                'Data copy time was {} that is {:.1f}% of a total time of {}. '
```
Comment (Tosa95) [inline pyungo/core.py:422]: Suggesting the user to disable deepcopy

**Review by cedricleroy — COMMENTED** [review]
Thanks for the PR @Tosa95! That's an interesting use case you have here.

The `deepcopy` is here to guarantee inputs immutability, as there is no way to do that properly with Python. To make this crystal clear, the example below would fail:
```python
graph = Graph(do_deepcopy=False)

@graph.register()
def f(c, e):
    c['a'] += 1
    c = c['a'] + e
    return f

d = {'a': 1}
res = graph.calculate(data={'c': d, 'e': 2})
assert res == 4
res = graph.calculate(data={'c': d, 'e': 2})
assert res == 4
```

But, as you noticed, there is a cost to pay using `deepcopy`.

I kinda like the idea, it gives responsibility to the user, and since it is not by default, it looks reasonable.

Do you mind adding some docstrings in the `Graph`, fix linting and maybe add one or two tests to validate the feature? Then it will be good to merge.

**Comment by Tosa95** [comment]
No doubts that `deepcopy` should be there by default.

I solved the problems you pointed out, I hope they are ok now.

Only a little question about immutability and side effects. If you run this code (I tested it):
```python
graph = Graph()

@graph.register()
def f():
    f = {'g': 1}
    return f

@graph.register(outputs=['h'])
def g(f):
    f['g'] += 1
    return f

@graph.register(outputs=['j'])
def i(f):
    f['g'] += 2
    return f

res = graph.calculate(data={})

print(graph.data.outputs)
```

You obtain:
`{'f': {'g': 4}, 'h': {'g': 4}, 'j': {'g': 4}}`

That shouldn't be considered correct.

The problem here is that the immutability should be kept during all the graph computation if we want to avoid this kind of side effects, so I suggest to add deep-copying at each computational step.

Obviously, it should be completely disabled if do_deepcopy is set to False.

**Comment by cedricleroy** [comment]
Thanks for the tests and the linting @Tosa95, looks good to me.

Regarding your example, this is correct, or at least as intended. Immutability is enforced in the inputs provided in the graph, but then the user is free to do whatever in the functions. If he/she wants to mutate, that's fine. If he/she wants to enforce immutability, this would need to happen in the functions. Hope it makes sense.

**Comment by Tosa95** [comment]
I think it's a matter of choice, I was saying that because most of the libraries involving DAG computation enforce immutability of objects (see Tensorflow or Apache Beam), however making the user capable of deciding is an extra degree of freedom so it's ok.

Thanks for this little library, it's very useful for small projects!

**Pull request merged**